### ISSUE TITLE
osh: 'trap' requires a code string

### ISSUE THREAD

**Issue by oguz-ismail** [oils-for-unix/oils#1881]
See:
```
$ for sh in bash bash-3.2.57 bosh 'busybox sh' dash gwsh ksh mksh oksh osh pdksh yash zsh; do
>   printf '%s: ' "$sh"
>   $sh -c 'trap : INT; trap'
> done
bash: trap -- ':' SIGINT
bash-3.2.57: trap -- ':' SIGINT
bosh: trap -- ':' INT
busybox sh: trap -- ':' INT
dash: trap -- ':' INT
gwsh: trap -- : INT
ksh: trap -- : INT
mksh: trap -- : INT
oksh: trap -- : INT
osh:   trap : INT; trap
              ^~~~
[ -c flag ]:1: 'trap' requires a code string
pdksh: trap -- : INT
yash: trap -- ':' INT
zsh: trap -- : INT
```
It'd be nice to have this basic feature in OSH (assuming it is actively developed).

**Comment by jasom** [comment]
I ran into this in some real-world code of mine; I have an "atexit" library I use, and as a sanity check, it greps for EXIT in the output of `trap`; this is not perfect, in that it can have false-positives, but I like that better than stomping on another EXIT handler.

**Comment by andychu** [comment]
(sorry for delayed response)

I am not sure how those 2 things are related?

Are you getting the error `'trap' requires a code string?`

I think we should fix it anyway, but this could be a separate issue

**Comment by jasom** [comment]
Yes, I'm getting the error `'trap' requires a code string` when running `trap` with no arguments in osh.

**Comment by oguz-ismail** [comment]
> I think we should fix it anyway

While you're at it fix this too
```
$ cat bug.sh
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
###################################################################################################
cc -xc - <<EOF
#include <unistd.h>
int main(void) {
    lseek(100, 15, SEEK_CUR);
}
EOF
./a.out
if false; then
        echo surprise!
        exit
fi
$
$ osh bug.sh
surprise!
$
```

**Comment by andychu** [comment]
- That doesn't relate to `trap`, so it should be a separate bug -- #2068 
- "Fix this too" is not how open source works -- **you** can fix it.  You can also fork the whole project
- If you want attention on this bug, then add some details about where you encountered it, or if it's a synthetic test case
  - that helps prioritize it among other work
  - synthetic test cases may still be fixed, depending on whether we think it leads to other bugs
  - I guess it's due to our use of descriptors above 100 for save/restore -- I think other shells may have similar variants of this issue, though I'm not sure.  Again, the bug needs context to be prioritized -- please respond on #2068 and not here

**Comment by oguz-ismail** [comment]
> I think other shells may have similar variants of this issue,

Other shells close internal file descriptors before execing a command.

> "Fix this too" is not how open source works

Don't care, I only responded because I received a notification. I don't test osh anymore

**Comment by andychu** [comment]
closed in favor of #2198
